server/src/test/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshotsTest.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.index.snapshots.blobstore;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.elasticsearch.repositories.blobstore.BlobStoreRepository.INDEX_SHARD_SNAPSHOTS_FORMAT;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+import org.elasticsearch.index.store.StoreFileMetadata;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Test;
+
+public class BlobStoreIndexShardSnapshotsTest extends ESTestCase {
+
+    @Test
+    public void test_bwc_streaming() throws Exception {
+        BlobStoreIndexShardSnapshots blobStoreIndexShardSnapshots = prepareData();
+        BytesStreamOutput out = new BytesStreamOutput();
+        blobStoreIndexShardSnapshots.writeTo(out);
+
+        StreamInput in = out.bytes().streamInput();
+        BlobStoreIndexShardSnapshots fromStream = BlobStoreIndexShardSnapshots.fromStream(in);
+        assertThat(isSame(blobStoreIndexShardSnapshots, fromStream)).isTrue();
+
+        out = new BytesStreamOutput();
+        out.setVersion(Version.V_5_10_1);
+        blobStoreIndexShardSnapshots.writeTo(out);
+
+        in = out.bytes().streamInput();
+        in.setVersion(org.elasticsearch.Version.V_5_10_1);
+        fromStream = BlobStoreIndexShardSnapshots.fromStream(in);
+        assertThat(isSame(blobStoreIndexShardSnapshots, fromStream)).isTrue();
+
+    }
+
+    @Test
+    public void test_number_of_unique_files_across_snapshots_equal_to_number_of_files() throws Exception {
+        BlobStoreIndexShardSnapshots blobStoreIndexShardSnapshots = prepareData();
+        BytesReference bytesReference = INDEX_SHARD_SNAPSHOTS_FORMAT.serialize(blobStoreIndexShardSnapshots, "dummyBlobName", true);
+
+        // FileInfo don't have equals/hashCode, so set will use identity check.
+        // That's exactly want we want:
+        // we need to make sure that we are using the same instance to save memory.
+        Set<FileInfo> uniqueWrittenFiles = new HashSet<>();
```

Comment:
the name `Written` is misleading, as we don't count anything during the `serialize`, we just take what is in the `BlobStoreIndexShardSnapshots` as built in `prepareData()`.

Reply:
Renamed the variable.

Actually, I can also check only` writeTo`, as @seut suggested in https://github.com/crate/crate/pull/17595#discussion_r1995297197 and don't tests reads (we can't add mock/spy for objects that are created by the testing method), no strong opinion here

Reply:
imho I prefer to leave it as is, since it tests what we want, (breaks without the fix), and we don't need to mock objects to achieve that. But let's hear @seut's opinion as well before merging.